lib/diff.mli

```diff
@@ -1,29 +1,33 @@
 type ('item, 'diff) entry =
   | Added of 'item
   | Removed of 'item
-  | Modified of 'diff
+  | Modified of 'diff  (** Represent a change of an entry in a collection *)
 
 type 'a atomic_modification = { reference : 'a; current : 'a }
+(** The simplest diff representation for the modification of a value of type 'a.
+    [reference] is the value before and [current] is the value after the change
+    occured. Use this type when there is no better representation available. *)
+
 type 'item atomic_entry = ('item, 'item atomic_modification) entry
 
 type ('item, 'diff) map = {
-  same : 'item String_map.t;
-  different : ('item, 'diff) entry String_map.t;
+  same_map : 'item String_map.t;
+  changed_map : ('item, 'diff) entry String_map.t;
 }
 
 type ('same, 'change) maybe_changed = Same of 'same | Changed of 'change
 
-type 'same atomic_maybe_changed =
+type 'same maybe_changed_atomic =
   ('same, 'same atomic_modification) maybe_changed
 
-type 'same atomic_maybe_changed_entry =
+type 'same maybe_changed_atomic_entry =
   ('same, 'same atomic_entry) maybe_changed
 
 type ('same, 'diff) option_ = ('same option, ('same, 'diff) entry) maybe_changed
 type 'same atomic_option = ('same, 'same atomic_modification) option_
 
 type ('same, 'diff) atomic_variant =
-  ('same, ('same atomic_maybe_changed, 'diff) maybe_changed) maybe_changed
+  ('same, ('same atomic_modification, 'diff) Either.t) maybe_changed
```

Comment:
Why don't use use a `maybe_changed` here?
`Same` would mean "same constructor with different argument values" and `Changed` would mean "different constructors".

Seeing `Left` and `Right` in the middle of code will not be self-explanatory on what they represent!

Reply:
I agree the `Either.t` is not the best suited type here but I don't think we should hide a diff in the `Same of x` argument of a `maybe_changed` as this is highly inconsistent with what it represent and likely going to confuse users.